src/main/java/org/traccar/model/User.java

```diff
@@ -273,4 +277,64 @@ public boolean isPasswordValid(String password) {
         return Hashing.validatePassword(password, hashedPassword, salt);
     }
 
+    private String totpKey;
+    private boolean useTotp;
+
+    public void setTotpKey(String key) {
+        if (key.isEmpty()) {
+            this.totpKey = null;
+            this.totpKey = this.getTotpKey();
+            return;
+        }
+        this.totpKey = key;
+    }
+
+    public String getTotpKey() {
+        if (this.totpKey == null) {
+            GoogleAuthenticator authenticator = new GoogleAuthenticator();
+            this.setTotpKey(authenticator.createCredentials().getKey());
+        }
+        return this.totpKey;
+    }
+
+    public void setUseTotp(boolean doUse) {
+        this.useTotp = doUse;
+    }
+
+    public boolean getUseTotp() {
+        if (!Context.getConfig().getBoolean("totp.enabled")) {
+            return false;
+        }
+        return this.useTotp;
+    }
+
+    @JsonIgnore
+    @QueryIgnore
+    public boolean isTotpAuthCodeValid(String code) {
```

Comment:
It would be nice to move most of the validation code our of this class.

Reply:
I can look if this can be moved to a new class.